src/OpenLoco/src/GameCommands/Track/RemoveTrack.cpp

```diff
@@ -0,0 +1,329 @@
+#include "RemoveTrack.h"
+#include "Economy/Economy.h"
+#include "Map/RoadElement.h"
+#include "Map/SignalElement.h"
+#include "Map/TileManager.h"
+#include "Map/Track/TrackData.h"
+#include "Map/TrackElement.h"
+#include "Objects/BridgeObject.h"
+#include "Objects/TrackExtraObject.h"
+#include "Objects/TrackObject.h"
+#include "Random.h"
+#include "RemoveSignal.h"
+#include "RemoveTrainStation.h"
+#include "Vehicles/Vehicle.h"
+#include "ViewportManager.h"
+
+namespace OpenLoco::GameCommands
+{
+    static World::TrackElement* getElTrackAt(const TrackRemovalArgs& args, const uint8_t trackFlags, const World::Pos3 pos, const uint8_t sequenceIndex)
+    {
+        auto tile = World::TileManager::get(pos);
+        for (auto& el : tile)
+        {
+            auto* elTrack = el.as<World::TrackElement>();
+            if (elTrack == nullptr)
+            {
+                continue;
+            }
+            if (elTrack->rotation() != args.rotation)
+            {
+                continue;
+            }
+            if (elTrack->sequenceIndex() != sequenceIndex)
+            {
+                continue;
+            }
+            if (elTrack->trackObjectId() != args.trackObjectId)
+            {
+                continue;
+            }
+            if ((elTrack->flags() & 0x30) != trackFlags)
+            {
+                continue;
+            }
+            if ((trackFlags & 0x10) != 0 && elTrack->owner() != getUpdatingCompanyId())
+            {
+                continue;
+            }
+            return elTrack;
+        }
+        return nullptr;
+    };
+
+    // 0x0049CC23
+    static currency32_t trackRemoveCost(const TrackRemovalArgs& args, const World::TrackData::PreviewTrack trackPiece0, const World::Pos3 trackStart, const uint8_t trackFlags)
+    {
+        const auto trackLoc = trackStart + World::Pos3{ Math::Vector::rotate(World::Pos2{ trackPiece0.x, trackPiece0.y }, args.rotation), trackPiece0.z };
+        auto* pieceElTrack = getElTrackAt(args, trackFlags, trackLoc, trackPiece0.index);
+        if (pieceElTrack == nullptr)
+        {
+            return 0;
+        }
+
+        currency32_t totalCost = 0;
+
+        const auto* trackObj = ObjectManager::get<TrackObject>(pieceElTrack->trackObjectId());
+        totalCost += Economy::getInflationAdjustedCost(trackObj->sellCostFactor, trackObj->costIndex, 10);
+
+        // Check mod removal costs
+        for (auto i = 0U; i < 4; i++)
+        {
+            if (pieceElTrack->hasMod(i))
+            {
+                const auto* trackExtraObj = ObjectManager::get<TrackExtraObject>(trackObj->mods[i]);
+                totalCost += Economy::getInflationAdjustedCost(trackExtraObj->sellCostFactor, trackExtraObj->costIndex, 10);
+            }
+        }
+
+        return totalCost;
+    }
+
+    // 0x004795D1
+    static void setLevelCrossingFlags(const World::Pos3 pos)
+    {
+        int8_t dh = (1 << 1) | (1 << 0);
+
+        while (true)
```

Comment:
Lets split this into two parts rather than use a while(true) that only iterates twice.

```cpp
auto findLevelTrackAndRoad = [](auto&& trackFunction, auto&& roadFunction){
    auto tile = World::TileManager::get(pos);
    for (auto& el : tile)
    {
        if (el.baseZ() != pos.z){continue;}
        if (el.isAiAllocated()){continue;}
        auto* elTrack = el.as<TrackElement>();
        if (elTrack != nullptr){
            if (elTrack->trackId() == 0){
               trackFunction(*elTrack);
            }
        }
        auto* elRoad = el.as<RoadElement>();
        if (elRoad != nullptr){
            if (elRoad->roadd() == 0){
               roadFunction(*elRoad);
            }
        }
    }
};
bool hasRoad = false;
bool hasTrack = false;
findLevelTrackAndRoad([&hasTrack](TrackElement& elTrack){ if (elTrack.hasLevelCrossing()){ hasTrack = true;}}, [&hasRoad](RoadElement& elRoad){ if (elRoad.hasLevelCrossing()){ hasRoad = true; }});
if (hasRoad ^ hasTrack)
{
    findLevelTrackAndRoad([hasTrack](TrackElement& elTrack){ if (hasTrack){ elTrack.setHasLevelCrossing(false);}}, [hasRoad](RoadElement& elRoad){ if (hasRoad){ elRoad.setHasLevelCrossing(false); elRoad.setUnk7_10(false); elRoad.setLevelCrossingObjectId(0); }});
}